dsfhub/resource_common.go

```diff
@@ -469,6 +471,109 @@ func auditStateRefreshFunc(client Client, resourceType string, assetId string) r
 	}
 }
 
+// waitForRemoteSyncState reads an asset periodically to check the status of remoteSyncState becomes "SYNCED"
+// posible values for remoteSyncState = ["SYNCED", "NOT_SYNCED", "UNKNOWN"]
+func waitForRemoteSyncState(ctx context.Context, resourceType string, assetId string, m interface{}) error {
+	client := m.(*Client)
+
+	stateChangeConf := &retry.StateChangeConf{
+		Pending: []string{
+			"NOT_SYNCED",
+			"UNKNOWN",
+		},
+		Target: []string{
+			"SYNCED",
+		},
+		Refresh:    remoteSyncStateRefreshFunc(*client, resourceType, assetId),
+		Timeout:    5 * time.Minute,
+		Delay:      10 * time.Second,
+		MinTimeout: 5 * time.Second,
+	}
+
+	_, err := stateChangeConf.WaitForStateContext(ctx)
+	if err != nil {
+		log.Printf("[ERROR] error while waiting for remoteSyncState = \"SYNCED\" for asset %v", assetId)
+		return err
+	}
+
+	return nil
+}
+
+// remoteSyncStateRefreshFunc reads an asset to check the status of remoteSyncState
+func remoteSyncStateRefreshFunc(client Client, resourceType string, assetId string) retry.StateRefreshFunc {
+	return func() (any, string, error) {
+		var result *ResourceWrapper
+		var err error
+
+		switch resourceType {
+		case dsfDataSourceResourceType:
+			{
+				log.Printf("[INFO] checking remote sync state for data_source asset %v", assetId)
+				result, err = client.ReadDSFDataSource(assetId)
+			}
+		case dsfLogAggregatorResourceType:
+			{
+				log.Printf("[INFO] checking remote sync state for log_aggregator asset %v", assetId)
+				result, err = client.ReadLogAggregator(assetId)
+			}
+		case dsfCloudAccountResourceType:
+			{
+				log.Printf("[INFO] checking remote sync state for cloud_account asset %v", assetId)
+				result, err = client.ReadSecretManager(assetId)
+			}
+		case dsfSecretManagerResourceType:
+			{
+				log.Printf("[INFO] checking remote sync state for secret_manager asset %v", assetId)
+				result, err = client.ReadLogAggregator(assetId)
+			}
+
+		default:
+			{
+				return nil, "", fmt.Errorf("invalid resourceType: %v", resourceType)
+			}
+		}
+		if err != nil {
+			return 0, "", err
+		}
+
+		return result, result.Data.RemoteSyncState, nil
+	}
+}
+
+// checkAuditState reads an asset to check the status of audit_pull_enabled
+func checkAuditState(ctx context.Context, m interface{}, assetId string, resourceType string) (bool, error) {
```

Comment:
it's not necessary whatsoever but it might be nice to extract the content of the switch statement into a helper since we are doing essentially the same thing three times now